### DISCUSSION TITLE
Github Action Static Linking

### DISCUSSION THREAD

**Discussion by nreinicke** [pykeio/ort#127] [Q&A] [answered]
I'm opening this as a discussion question since I don't think it's an issue with the library but rather my application of it.

I'm attempting to statically link in the onnx runtime from https://github.com/supertone-inc/onnxruntime-build/releases/. I've set up a [build script](https://github.com/NREL/routee-compass/blob/ndr/download-onnx-lib/rust/routee-compass-powertrain/build.rs) to download the appropriate lib at compile time and then set ORT_STRATEGY and ORT_LIB_LOCATION environment variables. This strategy is working when I test locally on MacOS (Monterey, Intel chip) and Windows 11 but when trying to use this on a [Github Action ](https://github.com/NREL/routee-compass/actions/runs/7133344825/job/19425862151)with ubuntu-latest, I'm getting a whole host of `undefined reference to` errors, making it seem like it can't find the c++ standard library.

I ran into similar errors when testing locally on a machine running Ubuntu 20.04 and was able to get it to compile by installing `cxx-compiler` from Anaconda (`conda install -c conda-forge cxx-compiler`) prior to building. But, when trying the same approach from Github Actions, I still hit these errors.

Has anyone run into similar issues? I'm not very familiar with c++ linking and so I might be missing something obvious but thought I would put this out in case anyone had a quick idea.

Thanks!

**Comment by decahedron1** [comment, marked as answer]
Supertone's static libraries are compiled on Ubuntu 22.04 so it's indeed very strange that your action isn't working with `ubuntu-latest`.

What I think may be happening is, for whatever reason, `ort` is not properly linking to `libstdc++`, but there is another dependency compiled before `ort` that does so *only on macOS*, hence why it works on your local macOS machine.

I see you're using ort v1.15, would it be possible to update to ort `2.0.0-alpha.2`? We have a migration guide here: https://ort.pyke.io/migrating/v2
v2 now provides static libraries by default, and with it there were also a few fixes to linking, specifically with `libstdc++`. Our CI runs are [working with stock `ubuntu-latest`](https://github.com/pykeio/ort/actions/runs/7108754900/job/19352586098) + static linking.

**Reply by nreinicke** [reply]
Thanks for your quick response! I've gone ahead and attempted to migrate to `2.0.0-alpha.2` (2.0 is much cleaner and I love that the static libraries are included by default!) and the action is passing!

I am hitting an error when running 2.0 locally on my Mac and on a linux server in the same network. 

```
[ort] failed to download https://parcel.pyke.io/v2/delivery/ortrs/packages/msort-binary/1.16.3/ortrs-msort_static-v1.16.3-x86_64-unknown-linux-gnu.tgz: Transport(Transport { kind: ConnectionFailed, message: Some("tls connection init failed"), url: Some(Url { scheme: "https", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("parcel.pyke.io")), port: None, path: "/v2/delivery/ortrs/packages/msort-binary/1.16.3/ortrs-msort_static-v1.16.3-x86_64-unknown-linux-gnu.tgz", query: None, fragment: None }), source: Some(Custom { kind: InvalidData, error: InvalidCertificate(UnknownIssuer) }) })
```

Given that the download is working on the action runner and not locally suggests that it's a problem internal to our network. Have you seen a similar error before for users trying to download binaries?

**Reply by decahedron1** [reply]
> Have you seen a similar error before for users trying to download binaries?

Hmm, no I haven't heard of any issues yet. The `InvalidCertificate(UnknownIssuer)` part is strange - the certificate is issued by Cloudflare so I'm not sure what could be causing that error. I have heard of some networks blocking domains under certain registrars, Namecheap (the registrar for `pyke.io`) being one of them, maybe that's related?

**Reply by nreinicke** [reply]
Yeah it's definitely something happening in our network as everything works outside of it, we'll just have to troubleshoot that. Thanks again for the quick help and for making this package available!
